src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -8,316 +8,269 @@
-    shared_sum = local_sum;
+  // Handle empty (trivial) input
+  if (l_n_features == 0 && r_n_features == 0) {
+    dh::LaunchN(out_splits.size(), [=]__device__(std::size_t idx) {
+      out_splits[idx] = DeviceSplitCandidate{};
+    });
+    return;
   }
-  cub::CTA_SYNC();
-  return shared_sum;
-}
 
-template <typename GradientSumT, typename TempStorageT> struct OneHotBin {
-  GradientSumT __device__ operator()(
-      bool thread_active, uint32_t scan_begin,
-      SumCallbackOp<GradientSumT>*,
-      GradientSumT const &missing,
-      EvaluateSplitInputs<GradientSumT> const &inputs, TempStorageT *) {
-    GradientSumT bin = thread_active
-                           ? inputs.gradient_histogram[scan_begin + threadIdx.x]
-                           : GradientSumT();
-    auto rest = inputs.parent_sum - bin - missing;
-    return rest;
+  auto out_scan = EvaluateSplitsGenerateSplitCandidatesViaScan(evaluator, left, right);
+  auto d_out_scan = dh::ToSpan(out_scan);
+
+  auto reduce_key = dh::MakeTransformIterator<int>(
+      thrust::make_counting_iterator<std::size_t>(0),
+      [d_out_scan] __device__(std::size_t i) {
+        return d_out_scan[i].node_idx;
+      });
+  auto reduce_val = dh::MakeTransformIterator<DeviceSplitCandidate>(
+      thrust::make_counting_iterator<std::size_t>(0),
+      [d_out_scan] __device__(std::size_t i) {
+        const auto& e = d_out_scan[i];
+        GradientSumT left_sum, right_sum;
+        if (e.is_cat) {
+          left_sum = e.parent_sum - e.partial_sum;
+          right_sum = e.partial_sum;
+        } else {
+          if (e.direction == DefaultDirection::kRightDir) {
+            left_sum = e.partial_sum;
+            right_sum = e.parent_sum - e.partial_sum;
+          } else {
+            left_sum = e.parent_sum - e.partial_sum;
+            right_sum = e.partial_sum;
+          }
+        }
+        return DeviceSplitCandidate{e.loss_chg, e.direction, e.findex, e.fvalue, e.is_cat,
+                                    GradientPair{left_sum}, GradientPair{right_sum}};
+      });
+  /**
+   * Perform segmented reduce to find the best split candidate per node.
+   * Note that there will be THREE segments:
+   * [segment for left child node] [segment for right child node] [segment for left child node]
+   * This is due to how we perform forward and backward passes over the gradient histogram.
+   */
+  dh::device_vector<DeviceSplitCandidate> out_reduce(3);
+  GPUTrainingParam param = left.param;
+  thrust::reduce_by_key(
+      thrust::device, reduce_key, reduce_key + static_cast<std::ptrdiff_t>(out_scan.size()),
+      reduce_val, thrust::make_discard_iterator(), out_reduce.data(),
+      thrust::equal_to<int>{},
+      [param] __device__(DeviceSplitCandidate l, DeviceSplitCandidate r) {
+        l.Update(r, param);
+        return l;
+      });
+  if (right.gradient_histogram.empty()) {
+    dh::LaunchN(1, [out_reduce = dh::ToSpan(out_reduce), out_splits]__device__(std::size_t) {
```

Comment:
Maybe we can fuse this kernel with the one in here https://github.com/dmlc/xgboost/blob/d27a427dc5e7a492b45f95c6bd99ea00a99b73b9/src/tree/updater_gpu_hist.cu#L342

Reply:
Maybe. It will make it harder to reason about `EvaluateSplits()` function however.

Reply:
From recent profiling results, small kernels have a significant impact when we grow deeper trees.  But I agree that it is premature optimization if we fuse them now.  Let's see the result first. ;-)